Build Issues with Cray Compilers for CUDA backend

### Describe the bug

Hi, 

I have had success in building the CUDA and HIP backend with Cray compilers in the past on SLES 15-SP5 OS. Using the following build script lead me to some build issues on [ALCF Polaris](https://www.alcf.anl.gov/polaris). It shouldn't matter though but tried with both Ninja and gmake builds and both pointed me to the same error. I am not sure if this has anything to do with GLIBC versions mismatch or GCC-toolchain path etc.


```bash
CUDA_LIB_PATH=/soft/compilers/cudatoolkit/cuda-12.6.1 python3 $DPCPP_HOME/llvm/buildbot/configure.py -o /soft/compilers/oneapi/upstream/install-${cuda_version}_${NOW} --cmake-opt="-DCMAKE_C_COMPILER=cc" --cmake-opt="-DCMAKE_CXX_COMPILER=CC" --native_cpu --cuda --llvm-external-projects openmp
cd /soft/compilers/oneapi/upstream/install-${cuda_version}_${NOW}
ninja sycl-toolchain
```

Error:
```
/usr/bin/ld: /usr/lib64/gcc/x86_64-suse-linux/12/../../../../lib64/Scrt1.o: in function `_start':
/home/abuild/rpmbuild/BUILD/glibc-2.31/csu/../sysdeps/x86_64/start.S:104:(.text+0x20): undefined reference to `main'

```
Verbose output:
```bash
: && /opt/cray/pe/craype/2.7.30/bin/CC -fPIC -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-maybe-uninitialized -Wno-nonnull -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wno-misleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -Wno-error -O3 -DNDEBUG  -Wl,-z,defs -Wl,-z,nodelete   -Wl,-z,relro,-z,now,-z,noexecstack -pie -Wl,--version-script=/soft/compilers/oneapi/upstream/install-12.6.1_01162025/_deps/unified-runtime-build/source/adapters/opencl/ur_adapter_opencl.map -shared -Wl,-soname,libur_adapter_opencl.so.0 -o lib/libur_adapter_opencl.so.0.12.0 _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/ur_interface_loader.cpp.o _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/adapter.cpp.o _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/command_buffer.cpp.o _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/common.cpp.o _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/context.cpp.o _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/device.cpp.o _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/enqueue.cpp.o _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/enqueue_native.cpp.o _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/event.cpp.o _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/image.cpp.o _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/kernel.cpp.o _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/memory.cpp.o _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/physical_mem.cpp.o _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/platform.cpp.o _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/program.cpp.o _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/queue.cpp.o _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/sampler.cpp.o _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/tensor_map.cpp.o _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/usm.cpp.o _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/usm_p2p.cpp.o _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/virtual_mem.cpp.o _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/__/__/__/__/unified-runtime-src/source/ur/ur.cpp.o _deps/unified-runtime-build/source/adapters/opencl/CMakeFiles/ur_adapter_opencl.dir/__/__/__/__/unified-runtime-src/source/common/umf_pools/disjoint_pool_config_parser.cpp.o  -Wl,-rpath,/soft/compilers/oneapi/upstream/install-12.6.1_01162025/lib:  -lstdc++fs  lib/libur_common.a  lib/libOpenCL.so.1.0.0  lib/libdisjoint_pool.a  lib/libumf.so.0.10.1  -lstdc++fs  -lrt  -ldl && :
/usr/bin/ld: /usr/lib64/gcc/x86_64-suse-linux/12/../../../../lib64/Scrt1.o: in function `_start':
/home/abuild/rpmbuild/BUILD/glibc-2.31/csu/../sysdeps/x86_64/start.S:104:(.text+0x20): undefined reference to `main'
```

Compiler versions:
```bash
/opt/cray/pe/craype/2.7.30/bin/CC --version
g++-12 (SUSE Linux) 12.3.0
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

/opt/cray/pe/craype/2.7.30/bin/cc --version
gcc-12 (SUSE Linux) 12.3.0
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

### To reproduce

_No response_

### Environment

- OS: SLES 15-SP5
- Target device and vendor: Nvidia A100, CUDA 12.6.1
- DPC++ version: [e.g. commit hash or output of `clang++ --version`]
- Dependencies version: [e.g. the output of `sycl-ls --verbose`]


### Additional context

_No response_

--- a/unified-runtime/cmake/helpers.cmake
+++ b/unified-runtime/cmake/helpers.cmake
@@ -157,11 +157,6 @@ function(add_ur_target_link_options name)
             if (UR_DEVELOPER_MODE)
                 target_link_options(${name} PRIVATE -Werror -Wextra)
             endif()
-            if (CMAKE_BUILD_TYPE STREQUAL "Release")
-                target_link_options(${name} PRIVATE
-                    $<$<CXX_COMPILER_ID:GNU>:-pie>
-                )
-            endif()
         endif()
     elseif(MSVC)
         target_link_options(${name} PRIVATE
@@ -176,7 +171,15 @@ function(add_ur_target_link_options name)
 endfunction()
 
 function(add_ur_target_exec_options name)
-    if(MSVC)
+    if(NOT MSVC)
+        if(NOT APPLE)
+            if(CMAKE_BUILD_TYPE STREQUAL "Release")
+                target_link_options(${name} PRIVATE
+                    $<$<CXX_COMPILER_ID:GNU>:-pie>
+                )
+            endif()
+        endif()
+    elseif(MSVC)
         target_link_options(${name} PRIVATE
             LINKER:/ALLOWISOLATION
         )

